addons/sale/models/sale_order_line.py

```diff
@@ -561,11 +561,19 @@ def _prepare_invoice_line(self, **optional_values):
             'discount': self.discount,
             'price_unit': self.price_unit,
             'tax_ids': [(6, 0, self.tax_id.ids)],
-            'analytic_tag_ids': [(6, 0, self.analytic_tag_ids.ids)],
             'sale_line_ids': [(4, self.id)],
         }
         if self.order_id.analytic_account_id and not self.display_type:
             res['analytic_account_id'] = self.order_id.analytic_account_id.id
+        has_account_rule = self.env['account.analytic.default'].sudo().account_get(
+                    product_id=self.product_id.id,
+                    partner_id=self.order_id.partner_id.id,
+                    user_id=self.env.uid,
+                    date=self.order_id.date_order,
+                    company_id=self.company_id.id,
+                )
+        if has_account_rule.analytic_tag_ids or self.analytic_tag_ids.ids:
+            res['analytic_tag_ids'] = [(6, 0, self.analytic_tag_ids.ids)]
```

Comment:
Seems strange to me.
I don't know the analytic flows, but providing the tags only if there is a default account (or whatever lies behind the account_get call) is very strange.
Wouldn't it make more sense to provide the tags only if there are some tags on the line ?

Reply:
The intention here is to prevent override if the user purposely removed the tag or added a tag with no rule set.

- if there is a rule:
1) if `self.analytic_tag_ids.ids = [ids]` -> normal flow `res['analytic_tag_ids']` is set
2)  if `self.analytic_tag_ids.ids = []` -> the user removed it on purpose => `res['analytic_tag_ids']` is set in order to not be overridden in account_move

- if there is no rule but there is tags on the line, it means that the user added a manual tag, and we don't want it to be overridden -> `res['analytic_tag_ids']` set with the manual tags
- if there is no rule and no tags, we don't set `res['analytic_tag_ids']` so if there is a rule (set with `account_id`), it can be applied in `account_move`  (thanks to the compute)

Reply:
Well, it seems that your logic relies on the fact that the analytic defaults give the same result between the SO creation and its invoicing, which might not be guaranteed.

This seems to be the kind of change that might fix the behavior for some clients while creating tickets from other clients bc of the modified flow :thinking:.

Anyway, as i don't really know the analytic logic, I'll leave the final decision & review to the accounting team.

Reply:
Yep, ok I'll see with them then.

Thanks anyway for the review

Reply:
To me, both `analytic_account_id` & should follow the same logic `analytic_tag_ids` so I'd like to suggest:
```py
        if self.order_id.analytic_account_id and not self.display_type:
            res['analytic_account_id'] = self.order_id.analytic_account_id.id
        if self.order_id.analytic_tag_ids and not self.display_type:
            res['analytic_tag_ids'] = [(6, 0, self.analytic_tag_ids.ids)]
```